### PR TITLE
H-3225: Allow querying of closed data schemas

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Closed data types are stored in the database but the graph currently cannot query them.
## 🔍 What does this change?

- After quite a bit of back and forth (thanks to @indietyp for the discussion about RTNs!) I moved out basic lookup of data types into a trait which we can improve later. (Future improvements covers building a caching layer around the trait or providing a proper in-memory store)
- This currently uses the `trait-variant` crate for simplicity (the issues we encountered don't apply in this PR).
- Implement (fairly simple) querying of closed data type schemas

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

### CODE REVIEW

File: libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+use type_system::{
+    schema::{ClosedDataType, ConversionExpression, DataTypeReference, DataTypeUuid},
+    url::BaseUrl,
+};
+
+use crate::ontology::DataTypeWithMetadata;
+
+#[trait_variant::make(Send)]
+pub trait DataTypeLookup {
+    type DataTypeWithMetadata: Borrow<DataTypeWithMetadata> + Send;
```
Comment (indietyp) [inline libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs:13]: Are you using `Borrow` here for the additional guarantees, or would `AsRef` be sufficient here?

**Review by indietyp — CHANGES_REQUESTED** [review]
one question, otherwise LGTM!

File: libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs (hunk 1 above)
Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs:13]: `AsRef` would probably be sufficient. But as `Borrow` has stronger guarantees (at least semantically) I chose that. Do you think I should change it?

Comment (TimDiekmann) [reply libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs:13]: Oh, and we use `Borrow` in other locations as well 😅

Comment (indietyp) [reply libs/@local/hash-graph-types/rust/src/ontology/data_type/lookup.rs:13]: I just wanted to make sure that you intended that those stronger guarantees are upheld^^

**Review by indietyp — APPROVED** [review]

**Review by indietyp — APPROVED** [review]

**Pull request merged**